src/main/scala/com/microsoft/hyperspace/index/CoveringIndex.scala

```diff
@@ -0,0 +1,420 @@
+ * for enabling builder pattern for [[CoveringIndexConfig]].
+ */
+object CoveringIndexConfig {
+
+  /**
+   * Builder for [[CoveringIndexConfig]].
+   */
+  class Builder {
+
+    private[this] var indexedColumns: Seq[String] = Nil
+    private[this] var includedColumns: Seq[String] = Nil
+    private[this] var indexName: String = ""
+
+    /**
+     * Updates index name for [[CoveringIndexConfig]].
+     *
+     * @param indexName index name for the [[CoveringIndexConfig]].
+     * @return an [[CoveringIndexConfig.Builder]] object with updated index name.
+     */
+    def indexName(indexName: String): Builder = {
+      if (!this.indexName.isEmpty) {
+        throw new UnsupportedOperationException("Index name is already set.")
+      }
+
+      if (indexName.isEmpty) {
+        throw new IllegalArgumentException("Empty index name is not allowed.")
+      }
+
+      this.indexName = indexName
+      this
+    }
+
+    /**
+     * Updates column names for [[CoveringIndexConfig]].
+     *
+     * Note: API signature supports passing one or more argument.
+     *
+     * @param indexedColumn indexed column for the [[CoveringIndexConfig]].
+     * @param indexedColumns indexed columns for the [[CoveringIndexConfig]].
+     * @return an [[CoveringIndexConfig.Builder]] object with updated indexed columns.
+     */
+    def indexBy(indexedColumn: String, indexedColumns: String*): Builder = {
+      if (this.indexedColumns.nonEmpty) {
+        throw new UnsupportedOperationException("Indexed columns are already set.")
+      }
+
+      this.indexedColumns = indexedColumn +: indexedColumns
+      this
+    }
+
+    /**
+     * Updates included columns for [[CoveringIndexConfig]].
+     *
+     * Note: API signature supports passing one or more argument.
+     *
+     * @param includedColumn included column for [[CoveringIndexConfig]].
+     * @param includedColumns included columns for [[CoveringIndexConfig]].
+     * @return an [[CoveringIndexConfig.Builder]] object with updated included columns.
+     */
+    def include(includedColumn: String, includedColumns: String*): Builder = {
+      if (this.includedColumns.nonEmpty) {
+        throw new UnsupportedOperationException("Included columns are already set.")
+      }
+
+      this.includedColumns = includedColumn +: includedColumns
+      this
+    }
+
+    /**
+     * Creates IndexConfig from supplied index name, indexed columns and included columns
+     * to [[CoveringIndexConfig.Builder]].
+     *
+     * @return an [[CoveringIndexConfig]] object.
+     */
+    def create(): CoveringIndexConfig = {
+      CoveringIndexConfig(indexName, indexedColumns, includedColumns)
+    }
+  }
+
+  /**
+   *  Creates new [[CoveringIndexConfig.Builder]] for constructing an [[CoveringIndexConfig]].
+   *
+   * @return an [[CoveringIndexConfig.Builder]] object.
+   */
+  def builder(): Builder = new Builder
+}
+
+case class CoveringIndexMetadata(
```

Comment:
This is a separate concern and should be separated in a separate file.

Reply:
Done